core/src/polling_monitor/mod.rs

```diff
@@ -0,0 +1,232 @@
+                    Err((id, e)) => {
+                        debug!(logger, "error polling";
+                                    "error" => format!("{:#}", e),
+                                    "object_id" => id.to_string());
+                        queue.lock().push_back(id)
+                    }
+                }
+            }
+        });
+    }
+
+    PollingMonitor {
+        queue,
+        wake_up_queue,
+    }
+}
+
+/// Handle for adding objects to be monitored.
+pub struct PollingMonitor<ID> {
+    queue: Arc<Mutex<VecDeque<ID>>>,
+
+    // This serves two purposes, to wake up the monitor when an item arrives on an empty queue, and
+    // to stop the montior task when this handle is dropped.
+    wake_up_queue: watch::Sender<()>,
+}
+
+impl<ID> PollingMonitor<ID> {
+    /// Add an object id to the polling queue. New requests have priority and are pushed to the
+    /// front of the queue.
+    pub fn monitor(&self, id: ID) {
+        let mut queue = self.queue.lock();
+        if queue.is_empty() {
+            // If the send fails, the response receiver has been dropped, so this handle is useless.
+            let _ = self.wake_up_queue.send(());
+        }
+        queue.push_front(id);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use anyhow::anyhow;
+    use futures::{Future, FutureExt, TryFutureExt};
+    use graph::log;
+    use std::{pin::Pin, task::Poll};
+    use tower_test::mock;
+
+    use super::*;
+
+    struct MockService(mock::Mock<&'static str, Option<&'static str>>);
+
+    impl Service<&'static str> for MockService {
+        type Response = (&'static str, Option<&'static str>);
+
+        type Error = (&'static str, anyhow::Error);
+
+        type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+        fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+            self.0.poll_ready(cx).map_err(|_| unreachable!())
+        }
+
+        fn call(&mut self, req: &'static str) -> Self::Future {
+            self.0
+                .call(req)
+                .map_ok(move |x| (req, x))
+                .map_err(move |e| (req, anyhow!(e.to_string())))
+                .boxed()
+        }
+    }
+
+    async fn send_response<T, U>(handle: &mut mock::Handle<T, U>, res: U) {
+        handle.next_request().await.unwrap().1.send_response(res)
+    }
+
+    #[tokio::test]
+    async fn polling_monitor_simple() {
+        let (svc, mut handle) = mock::pair();
+        let (tx, mut rx) = mpsc::channel(10);
+        let monitor = spawn_monitor(MockService(svc), tx, log::discard());
+
+        // Basic test, single file is immediately available.
+        monitor.monitor("req-0");
+        send_response(&mut handle, Some("res-0")).await;
+        assert_eq!(rx.recv().await, Some(("req-0", "res-0")));
+    }
+
+    #[tokio::test]
+    async fn polling_monitor_unordered() {
```

Comment:
I'm not sure I've followed how this tests the "unorderedness" of the PollingMonitor since the test does indeed send/receive in a specific order, can you explain it a bit further please?

Reply:
It sends in the order `req-0`, `req-1`, meaning `req-1` is called first on the service. But the response for `req-0`  arrives first and is received without waiting for `req-1`. This tests that we're using `.call_all(stream).unordered()`.